q2_dada2/_denoise.py

```diff
@@ -132,6 +135,15 @@ def _denoise_helper(biom_fp, track_fp, hashed_feature_ids):
     sid_map = {id_: _filepath_to_sample(id_)
                for id_ in table.ids(axis='sample')}
     table.update_ids(sid_map, axis='sample', inplace=True)
+    # Reintroduce empty samples dropped by dada2.
+    table_cols = table.ids(axis='observation')
+    table_rows = list(set(df.index) - set(table.ids()))
+    table_to_add = biom.Table(np.zeros((len(table_cols), len(table_rows))),
+                              table_cols, table_rows,
+                              type="OTU table")
+    table = table.concat(table_to_add)
+    if not retain_all_samples:
```

Comment:
Probably best to comment this to explain that there are other sources of empty samples, not just the 6ish lines above this (otherwise, you might be tempted to stash everything in an if/else statement).

Reply:
Good idea!  I inserted one.